### PR TITLE
Consistent addresses for upgradable contracts

### LINKED ISSUE #1350: Use consistent addresses for v2 mailboxes
The current create2 tooling does not support passing arguments to the constructor, nor does it support consistent addresses for upgradable contracts. The mailboxes have both of these

### PR DESCRIPTION
**Pull request by asaj**
### Description

This PR modifies the upgradability proxy pattern from `UpgradeBeacon` to `Transparent` in order to support creating upgradable contracts with consistent addresses across networks.

To create proxies at different addresses, a salt is generated from the environment and contract name. 

The actual process of deploying a proxy to a consistent address requires some constructor argument gymnastics, which are described in comments. 

### Drive-by changes

- Adds batch enrollment functions to MultisigIsm to make deployment faster
 
### Related issues

- Fixes #1350 

### Backward compatibility

No

### Testing

Unit Tests, manually deployed to testnet and saw consistent proxy addresses

### CODE REVIEW

File: typescript/sdk/src/proxy.ts [hunk 1]
```diff
@@ -27,10 +27,7 @@ export function isProxyAddresses(
   );
 }
 
-export interface BeaconProxyAddresses
-  extends ProxyAddresses<ProxyKind.UpgradeBeacon> {
-  beacon: types.Address;
-}
+export type TransparentProxyAddresses = ProxyAddresses<ProxyKind.Transparent>;
```
Comment (yorhodes) [inline typescript/sdk/src/proxy.ts:30]: would have expected the admin address here if its required everywhere
though I see how it could be optional

**Review by yorhodes — APPROVED** [review]

File: typescript/infra/src/core/deploy.ts [hunk 2]
```diff
@@ -12,9 +25,55 @@ import { writeJSON } from '../utils/utils';
...
+  async deployInterchainGasPaymaster<LocalChain extends Chain>(
+    chain: LocalChain,
+    proxyAdmin: ProxyAdmin,
+  ): Promise<
+    ProxiedContract<InterchainGasPaymaster, TransparentProxyAddresses>
+  > {
+    const deployOpts = {
+      create2Salt: ethers.utils.solidityKeccak256(
+        ['string', 'string'],
+        // TODO: Fix salt
```
[the remark this answers is not in the document]
Comment (asaj) [reply typescript/infra/src/core/deploy.ts:47]: The "1" suffix at the end

File: typescript/infra/src/core/deploy.ts [hunk 3]
```diff
@@ -12,9 +25,55 @@ import { writeJSON } from '../utils/utils';
...
+
+  async deployMailbox<LocalChain extends Chain>(
+    chain: LocalChain,
+    defaultIsmAddress: types.Address,
+    proxyAdmin: ProxyAdmin,
+  ): Promise<ProxiedContract<Mailbox, TransparentProxyAddresses>> {
+    const deployOpts = {
+      create2Salt: ethers.utils.solidityKeccak256(
+        ['string', 'string'],
+        // TODO: Fix salt
```
[the remark this answers is not in the document]
Comment (asaj) [reply typescript/infra/src/core/deploy.ts:62]: 1 suffix

File: typescript/sdk/src/deploy/HyperlaneDeployer.ts [hunk 4]
```diff
@@ -157,21 +160,28 @@ export abstract class HyperlaneDeployer<
...
+        ethers.utils.concat([factory.bytecode, encodedConstructorArgs]),
+      );
+
+      // TODO: Maybe recover deployed contracts?
       const contractAddr = await create2Factory.deployedAddress(
-        factory.bytecode,
+        bytecode,
         await signer.getAddress(),
         salt,
       );
```
[the remark this answers is not in the document]
Comment (asaj) [reply typescript/sdk/src/deploy/HyperlaneDeployer.ts:174]: Yeah it's not hard just not necessary yet

File: typescript/sdk/src/deploy/HyperlaneDeployer.ts [hunk 5]
```diff
@@ -234,31 +244,76 @@ export abstract class HyperlaneDeployer<
   protected async deployProxy<C extends ethers.Contract>(
     chain: Chain,
     implementation: C,
-    beaconAddress: string,
+    proxyAdmin: ProxyAdmin,
```
[the remark this answers is not in the document]
Comment (asaj) [reply typescript/sdk/src/deploy/HyperlaneDeployer.ts:247]: Abstract how? We need the `upgradeAndCall`

File: typescript/sdk/src/deploy/HyperlaneDeployer.ts [hunk 6]
```diff
@@ -285,61 +340,35 @@ export abstract class HyperlaneDeployer<
...
-  ): Promise<ProxiedContract<C, BeaconProxyAddresses>> {
-    this.logger(`Duplicate Proxy on ${chain}`);
-    return this.deployProxy(
-      chain,
-      proxy.contract.attach(proxy.addresses.implementation) as C,
-      proxy.addresses.beacon,
-      initArgs,
-    );
-  }
-
```
[the remark this answers is not in the document]
Comment (asaj) [reply typescript/sdk/src/deploy/HyperlaneDeployer.ts:342]: We could but my inclination is to remove unused code whenever we find it

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts [hunk 7]
```diff
@@ -66,39 +91,72 @@ export class HyperlaneCoreDeployer<
...
+        (entry): entry is [number, number, string[]] => entry[2] !== undefined,
+      );
+      if (validatorEntries.length > 0) {
+        const validatorDomains = validatorEntries.map(([id]) => id);
+        const validatorAddresses = validatorEntries.map(
+          ([, , addresses]) => addresses,
         );
-        const threshold = await multisigIsm.threshold(domain);
-        if (!threshold.eq(multisigIsmConfig.threshold)) {
+        for (const entry of validatorEntries) {
```
[the remark this answers is not in the document]
Comment (asaj) [reply typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts:125]: Which, the threshold equality? We do it above and don't create a threhsold entry if it's already set properly

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts [hunk 8]
```diff
@@ -66,39 +91,72 @@ export class HyperlaneCoreDeployer<
...
+      );
+      if (thresholdEntries.length > 0) {
+        const thresholdDomains = thresholdEntries.map(([id]) => id);
+        const thresholdChains = thresholdDomains.map(
+          (id) => DomainIdToChainName[id],
+        );
+        const thresholds = thresholdEntries.map(([, threshold]) => threshold);
+        this.logger(
+          `Set remote (${thresholdChains}) thresholds on ${chain}: ${thresholds}`,
+        );
```
[the remark this answers is not in the document]
Comment (asaj) [reply typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts:152]: Cleaned this up a bit

File: typescript/sdk/src/proxy.ts (hunk 1 above)
Comment (asaj) [reply typescript/sdk/src/proxy.ts:30]: It felt redundant given we had a separate entry for the proxy admin. I don't feel strongly

File: typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (yorhodes) [reply typescript/sdk/src/deploy/core/HyperlaneCoreDeployer.ts]: the `validatorEntries.length > 0`

**Pull request merged**